odoo/api.py

```diff
@@ -1457,6 +1457,37 @@ def process(model, field, field_cache):
         if invalids:
             _logger.warning("Invalid cache: %s", pformat(invalids))
 
+    def _get_grouped_context_dependent_field_cache(self, field, group_key=lambda ck: ck[0]):
+        """
+        get a field cache proxy to group up field value for all context
+        cache data: {field: {context_key: {id: value}}}
+
+        :param field: field for the grouped field cache value
+        :param group_key: a function to convert the context key tuple to a
+               group key. None group key will be ignored
+        :return: a dict like field cache proxy which is logically similar to
+              {id: {group_key(context_key), value}}
+        """
+        field_caches = self._data.get(field, EMPTY_DICT)
+        key_field_cache = {
+            key: field_cache
+            for context_key, field_cache in field_caches.items()
+            if (key := group_key(context_key)) is not None
+        }
+        return GroupedContextDependentFieldCache(key_field_cache)
+
+
+class GroupedContextDependentFieldCache:
+    def __init__(self, key_field_cache):
+        self._key_field_cache = key_field_cache
+
+    def __getitem__(self, id_):
+        return {
+            key: field_cache[id_]
+            for key, field_cache in self._key_field_cache.items()
+            if id_ in field_cache
+        }
```

Comment:
I see your point. IMO, it looks like early optimization, memory issue is for huge amount of record/field/data. IMO, it will not be an issue for company_depend fields (company_depend field is spares most of the case anyway).

As you want :smiley: